sat/apiclient/pcs.py

```diff
@@ -0,0 +1,189 @@
+#
+# Permission is hereby granted, free of charge, to any person obtaining a
+# copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+"""
+Basic client library for PCS.
+"""
+from collections import defaultdict
+
+from csm_api_client.service.gateway import APIError, APIGatewayClient
+from csm_api_client.service.hsm import HSMClient
+from sat.xname import XName
+
+
+class PCSError(APIError):
+    """An error occurred in PCS."""
+
+    def __init__(self, message, xname_errs=None):
+        """Create a new PCSError with the given message and info about the failing xnames.
+
+        Args:
+            message (str): the error message
+            xname_errs (list): a list of dictionaries representing the failures for
+                the individual components that failed. Each dict should have the
+                following keys:
+                    e: the error code
+                    err_msg: the error message
+                    xname: the actual xname which failed
+        """
+        self.message = message
+        self.xname_errs = xname_errs if xname_errs is not None else []
+        self.xnames = [xname_err['xname'] for xname_err in self.xname_errs
+                       if 'xname' in xname_err]
+
+    def __str__(self):
+        """Convert to str."""
+        if not self.xname_errs:
+            return self.message
+        else:
+            # A mapping from a tuple of (err_code, err_msg) to a list of xnames
+            # with that combination of err_code and err_msg.
+            xnames_by_err = defaultdict(list)
+            for xname_err in self.xname_errs:
+                xnames_by_err[(xname_err.get('e'), xname_err.get('err_msg'))].append(xname_err.get('xname'))
+
+            xname_err_summary = '\n'.join([f'xname(s) ({", ".join(xnames)}) failed with '
+                                           f'e={err_info[0]} and err_msg="{err_info[1]}"'
+                                           for err_info, xnames in xnames_by_err.items()])
+
+            return f'{self.message}\n{xname_err_summary}'
```

Comment:
All of the `xname_errs` handling does not apply to PCS and should be removed.

```suggestion
class PCSError(APIError):
    """An error occurred in PCS."""
```

Reply:
Make sure usages don't use `xname_errs` and update or remove test that uses it as well.

Reply:
xname_errs has been removed along with the tests the uses it